### PR TITLE
Create Templates-* host groups

### PR DESCRIPTION
**Pull request by pederhan**
This pull request makes the application create `Templates-*` host groups in addition to `Siteadmin-*` host groups. Additionally, the prefix used to create the templates groups can be configured via the new `hostgroup_templates_prefix` configuration option. 

Should the prefix in the mapping file ever change from `Siteadmin-` in the future, it must be changed accordingly in the config with the new `hostgroup_siteadmin_prefix` configuration option. We rely on this value to inform the program of how to construct the templates host group names from the siteadmin group names in the file.

## Primary changes

* Added functionality for automatically creating `Templates-*` host groups
* Added config option for adjusting the prefix used to create the templates host groups (default: `Templates-`)
* Added config option to specify what the currently used prefix for siteadmin groups is (default: `Siteadmin-`)
* Added tests for new utils functions for constructing host group names.

## Other changes

* Added missing logging calls for certain actions when run in dryrun mode.
* Fixed comment in `utils.py` regarding constrained types.

## Questions

Should there be a toggle for creating the `Templates-*` group, or do we make it mandatory for these groups to be created?

### CODE REVIEW

**Comment by rafaelma** [comment]
The creation/use of Siteadmin-* and Templates-* host groups was always intended to be used as containers that could be used to register hosts and templates owned by a particular user group. It is a way of organizing and categorizing both hosts and templates when you have many user groups using a Zabbix installation.

These host groups can also be used to assign privileges in Zabbix if you want. We use them for this in addition to the organization/categorization of the data, but this is handled by mechanics other than zac. We use other tools to define this access configuration.

The reason for using two type of host groups (Siteadmin-* and Templates-*) is because they contains different types of data, and again, they are used as a way of organizing and categorizing data.

Some open questions:

- Maybe we should change the default prefix for the "hosts" group from Siteadmin-<owner> to Hosts-<owner> to avoid confusions about what they contain. With the use of the two added config options to specify these prefixes, this is a simple task now.
- Actually the idea of using a switch to create or not the Siteadmin/Hosts-* and Templates-* groups is a good idea. Not everybody has the need of using these groups so we don't need to make them mandatory.

**Comment by rafaelma** [comment]
The arguments against the points you made are:
- `Templates-` groups and `Siteadmin-` groups main function is to  **organize** and **categorize** host and template data/objects. It is a way of knowing who has the responsability for a host or template. The same as we do with `Source-`, `Role-`, `Importance-`, `All-`, .... host groups, they are used to **organize** and **categorize** data/objects. This has always been done in zac and I don't see a point of moving this out of zac.
- That `Templates-` and `Siteadmin-` groups are used "in addition" to define what users can access or not, is not a good argument to move/duplicate the code used to create these two host groups in zac into another external  tool. Many of the other host groups are also used to define what users can access or not, are we also going to move the management of these host groups to another tool outside zac?
- The access/privileges and user management are not defined or handled in zac, we use another tool to implement this. That zabbix uses hostgroups to define permissions doesn't mean that we have to move the creation of host groups out of zac. Again all host groups are used to **organize** and **categorize** host and template data/objects.
- Templates are templates, and hosts are hosts, they are two different types of data/objects in Zabbix, why do you want to have two different types of data/objects in the same host group? Isn't this more confusing than having different host groups for templates and for hosts? Even Zabbix has per default a dedicated host group `Templates` for templates. Templates are also a part of host management, and zac takes care for example of linking/unlinking templates to host/devices.
- I have check our Zabbix installation. We have 82 `Templates-` host groups, none of them has a single host registered on it. We have 90 `Siteadmin-` hostgroups and 14 of them has some templates registered on it. These is wrong and we will fix it and inform our users.  
- To avoid confusions about the use of `Siteadmin-` host groups, we have the suggestion of changing the default prefix for the "hosts" group from `Siteadmin-` to `Hosts-` (or `Devices-`, to avoid confusion when printers, network boxes, AV devices, etc are also monitored) this will avoid confusions about what these host groups contain.

**Comment by paalbra** [comment]
> * `Templates-` groups and `Siteadmin-` groups main function is to  **organize** and **categorize** host and template data/objects. It is a way of knowing who has the responsability for a host or template. The same as we do with `Source-`, `Role-`, `Importance-`, `All-`, .... host groups, they are used to **organize** and **categorize** data/objects. This has always been done in zac and I don't see the point of moving this out of zac.

They are used to organize access to host and templates. The groups perform the same task: managing access. Zac doesn't do access management.

> * That `Templates-` and `Siteadmin-` groups are used "in addition" to define what users can access or not, is not a good argument to move/duplicate the code used to create these two host groups in zac into another external  tool. Many of the other host groups are also used to define what users can access or not, are we also going to move the management of these host groups to another tool outside zac?

Duplicate code? Zac doesn't do access management. Other tools do.

> * The access/privileges and user management are not defined or handled in zac, we use another tool to implement this. That zabbix uses hostgroups to define permissions doesn't mean that we have to move the creation of host groups out of zac. Again, all host groups are used to **organize** and **categorize** host and template data/objects.

Yes, configuration of access should be outside of zac. Since zac doesn't do access or user management.

> * Templates are templates, and hosts are hosts, they are two different types of data/objects in Zabbix, why do you want to have two different types of data/objects in the same host group? Isn't this more confusing than having different host groups for templates and for hosts? Even Zabbix has per default a dedicated host group `Templates` for templates. Templates are also a part of host management, and zac takes care for example of linking/unlinking templates to host/devices.

Templates and hosts is two objects that require granted access. A group that is used to grant this access could and probably should be a single group. Managing templates is not host management (linking templates is host management. It's performed on the host. Managing templates/access and linking templates is completely different).

> * I have check our Zabbix installation. We have 82 `Templates-` host groups, none of them has a single host registered on it. We have 90 `Siteadmin-` hostgroups and 14 of them has some templates registered on it. This is wrong and we will fix it and inform our users.

It could be fixed by merging the groups. This is my point.

> * To avoid confusions about the use of `Siteadmin-` host groups, we have the suggestion of changing the default prefix for the "hosts" group from `Siteadmin-` to `Hosts-` (or `Devices-`, to avoid confusion when printers, network boxes, AV devices, etc are also monitored) this will avoid confusions about what these host groups contain.

It could be fixed by merging the groups.

Bottom line:

This is not host management (template., user and access management is not in the current scope of zac). This PR should not be merged. This is a implicit and hacky design. @pederhan 's question is very legit, but it shouldn't be performed at all. Merging the two access groups is something that rather should be looked into.

**Comment by paalbra** [comment]
Looking into growing the scope of zac into user and access management could of course also be done. But these access groups could and should still be a single group.

**Comment by paalbra** [comment]
My main issue has been the proposed design change. I haven't looked at the actual code, nor have I run the code. Now I have.

It doesn't seem to work like the proposed design either. It has multiple issues.

The new prefix values in the config creates issues.

Also hosts are actually added to the `Template-` groups, which make the groups true duplicates.

Running zac like in the README.md "quick start" will result in something like:
image

But like I said. I don't think the proposed *design* should be implemented. The PR should not be merged (even if the above is fixed).

**Comment by paalbra** [comment]
The implementation still has issues:

image

Once more: This shouldn't be implemented. It's a step backwards. I can't believe I'm the only one that sees the issues here? Do you all agree that this is a good thing? Please, please point out the good parts here?

**Comment by pederhan** [comment]
Yes, that's due to the default config value `mapping_file_prefix` being `"Siteadmin"` instead of `"Hostgroup-"`, which is a mistake. I was testing with `"Siteadmin-"` in the example mapping file, and forgot the example file uses `"Hostgroup-"`. Here's how it looks on my end fwiw:

image

We are scrapping this pull request, but intend to submit a new PR that enables creation of additional host groups based on the names in the hosts map file, in a way that does not affect the existing functionality for managing host groups and their hosts.

**Comment by paalbra** [comment]
No worries. I see the confusion, since they are used (exclusively?) at UiO.

I would really (really, really) suggest against this functionality though. IMHO it adds no value at all. It's weird and implicit behavior, it's confusing for users, and it probably creates loads of empty host groups which are never used for anything (how is this status at UiO?).

Like I've mentioned before: Do not create `Siteadmin-` and `Templates-` groups at UiO. They should be merged. It simplifies the design. They perform the same task. It's no need to duplicate them.

I'm ready to be convinced though, if any good arguments/scenarios where this really is useful is explained :+1:

Also, long term: I think it's a good idea to look into expanding zac into user and permission management. It would make zac a more complete app (and would remove the not so good relation between different management systems at UiO).
